### PR TITLE
Allow errors converting custom types to propagate

### PR DESCRIPTION
**Pull request by mhammond**
This gets close, but strikes a mismatch between the `Err` types.

### CODE REVIEW

**Review by bendk — COMMENTED** [review]
The approach seems reasonable.  Updating that FFI calling code is going to be annoying, but it seems doable.

My one concern is that the error type of the wrapper needs to match the error type of the method.  In practice, I think this is going to mean that libraries need to define a global error type.  That's not the worst, especially if we implemented something like #1076, but it does seem like it constrains how the libraries need to be written.

I'd prefer if we could make it so the error types for the wrapper and methods weren't linked together.  It seems like this should be possible with some modification to the `rustcall` module.

**Comment by mhammond** [comment]
This is all green apart from the "uitests", and I'm very confused by them :) See cbd152c35fbf9d1f01be071853bfa37fd7d5ce8b - *some* changes are expected, but the change from `$DIR/counter.uniffi.rs:160:13` to `$WORKSPACE/target/debug/build/uniffi_uitests-6472345683e61af2/out/counter.uniffi.rs` is obviously a problem, but that's what was generated when I took the advice to use `TRYBUILD=overwrite` - anyone have any clue about this?

File: fixtures/ext-types/guid/src/guid.udl
Comment (mhammond) [inline fixtures/ext-types/guid/src/guid.udl:21]: wrong bracket! Should be `Err(GuidError)`

File: CHANGELOG.md
Comment (bendk) [inline CHANGELOG.md:15]: Nit: there's a extra `[` before "may"

File: fixtures/ext-types/guid/src/lib.rs
Comment (bendk) [inline fixtures/ext-types/guid/src/lib.rs:26]: IIUC if this assertion fails, then that would be a test error.  I think this string would make it hard to figure out what went wrong in that case though.  How about changing the it to "our UniffiCustomTypeWrapper failed to check for an empty GUID" or something like that.

File: docs/manual/src/udl/ext_types_wrapped.md
Comment (bendk) [inline docs/manual/src/udl/ext_types_wrapped.md:76]: I like how this lets us keep using `anyhow::Result` as the return type in all cases.

File: uniffi_bindgen/src/scaffolding/templates/macros.rs
Comment (bendk) [inline uniffi_bindgen/src/scaffolding/templates/macros.rs:33]: This code looks good to me.  Does it get inserted for each function argument though?  Seems like that could make the scaffolding really hard to read.  I wonder if it could be refactored into a function in `uniffi` somehow, but I don't think that should block merging this.

**Review by bendk — APPROVED** [review]
The more I think about the options, the more I think the least bad one is to require that the wrapper and function error types match.  Maybe we could relax that in the future, but maybe we won't need to either.

Given that approach, I like this implementation a lot.  Keep everything using `anyhow` and use `downcast_ref` to check the error type.

I think the only issue is the UI tests, no clue what's happening with those though.  I'll make approve, but I do think we should get those green before merging.

**Review by tarikeshaq — APPROVED** [review]
LGTM! Glad `anyhow` is making our lives easier 🕺 

I agree with Ben that we might want a long-term discussion/fix to mark the error story better, but this here seems like the best way forward for now!

(I haven't looked at the uitests before so also not sure I can give any guidance there :( )

**Comment by mhammond** [comment]
Thanks for the comments. I've still not tried to tackle records yet, but I think this is OK to land and unstick our "places" work.

**Pull request merged**
